### DISCUSSION TITLE
The experience of syncing Lightdash with dbt should change.

### DISCUSSION THREAD

**Discussion by TuringLovesDeathMetal** [lightdash/lightdash#1143] [Ideas]
### ❓ What questions do we have for our users/community?
- When do you use the `refresh dbt` button?
- When was the last time you got an error from clicking `refresh dbt`? What was it?
- When you've refreshed dbt, do you ever have to leave the Lightdash app to find information about the refresh? (debugging an error, seeing if it's actually syncing, etc.)
- Which step in making changes in dbt --> seeing them in Lightdash takes you the longest to complete right now?

### 💡 What’s the idea?
**Step 1: we automatically refresh dbt when there are changes in your GitHub connection**

The problem: the `refresh dbt` button is confusing. When do I press it? Why should I press it vs. running a query? We can remove this button to make it less confusing to users and make the connection to dbt way cleaner if we automatically refresh dbt for you when we detect changes in your GitHub project.

- We could make a CI/CD job - a GitHub action. When there's a merge to the branch your Lightdash project is connected to, it will compile your project and send this manifest.json to Lightdash.
- This removes the `refresh dbt` button.

**Step 2: we revamp the dbt connection page**
The problem: it's hard to debug dbt connection problems right now. Most of the errors that users run into come from their connection with their dbt project (both in onboarding + continuously). It should be easier to debug your Lightdash connection to dbt and understand the current state of your dbt project in Lightdash (when was it last synced? What commit has it synced to? What are the errors?)

- We create a page with information about the status of the dbt connection. 
- we tell you the commit that your Lightdash project is synced to
- We tell you when the most recent Lightdash sync was done (i.e. when your project was last re-compiled)
- You get the logs of the refresh coming in as it's happening and once it's completed. You could see a history of compiles and other details about your connection (à la [dbt cloud run history.](https://cloud.getdbt.com/#/accounts/20983/projects/34669/runs/40084552/) or dataform’s run history)
- We give you the errors + warnings from the most recent compile. We also give you actions to solve the most common ones:
  - do you have a dimension that was added to your dbt project, but we can't find it in the warehouse?

**Step 3: we check if the changes your making will break Lightdash.**

If you open a PR, then we check to see if this will work with your Lightdash project:

- we'd have a GitHub action that runs dbt compile
- it sends the manifest to your Lightdash instance - wouldn’t change what’s in Lightdash, but would compile manifest and send you back the state and tell you if you were going to break stuff in Lightdash.
- (we'd need to build an API with an endpoint that could test the changes in Lightdash)

### 📝 What’s the rough overview of work required to ship this?

*do we need user research? do we need to plan/partner with another company? do we need to find a new tool to do the thing we want to do?*

- user research with a prototype/early sketch of what we want in there (show some example errors + some example information you’d have in the compile page).
- design
- engineering

### ⏭ What are the next steps?

*If we started working on this tomorrow, what would we do first? If you decide to kill this idea, talk about your reasons in here.*

- decide on the initial information we think is important in this page + create an early prototype and test it with some users.

### 😎 Any other notes/comments/important info

- Make the flow from changing something in your dbt project —> seeing changes in Lightdash clearer. Here’s the steps we have right now:
    - I made a change in dbt
    - Will this change break anything in Lightdash?
    - I merge this to the branch that’s connected to Lightdash (normally `main`)
    - Is this change affecting metrics only? Is this affecting dimensions?
    - If it’s affecting dimensions, have you updated the data in your data warehouse?
    - Has your Lightdash project synced with your changes in GitHub?
    - If yes, you can now click `refresh dbt`

**Comment by owlas** [comment]
Looks great, thanks for distilling this down @TuringLovesDeathMetal !

**Comment by TuringLovesDeathMetal** [comment]
**When do you use the `refresh dbt` button?**
When I've made a change to my dbt project and merged it to GitHub and I want to sync the changes into Lightdash.

**When was the last time you got an error from clicking refresh dbt? What was it?**
Adding a new dimension to dbt but forgetting to `run dbt` so it wasn't in my warehouse yet. Got a `column not found in warehouse` or something like that error.

**When you've refreshed dbt, do you ever have to leave the Lightdash app to find information about the refresh? (debugging an error, seeing if it's actually syncing, etc.)**
- Yes. If there's an error with the dbt code, I have to go back to dbt to test it. 
- If I thought I made a change that fixes the problem, I go back and check GitHub to make sure the changes were merged. I'll also re-run `refresh dbt` in case it didn't actually refresh last time. 
- When I'm not sure if Lightdash had the most recent GitHub changes before it tried to re-sync (there's nowhere to go right now to see if this is true or not...but it's not in the Lightdash app either)

**Which step in making changes in dbt --> seeing them in Lightdash takes you the longest to complete right now?**
The step between merging a change in GitHub --> seeing that change in Lightdash. 

I'm never sure if Lightdash has synced to my most recent merge, so if I click `refresh dbt`, is it pulling the most recent changes? Or the changes before that because it's still rendering? 

**These are the other questions I run into when I'm refreshing dbt:**
- Does Lightdash have my latest change?
- Is the compile successful? If not, what are the errors?
- If any columns/tables have warnings, what are the errors? (I want to be able to see this in one place)
- If I've added a new dimension and forgot to run dbt to add it to the warehouse, I always get confused when there's an error from this.

**Reply by djayatillake** [reply]
If there hasn't been a github change could Lightdash persist what it gets from the button on startup? That way startup will be much faster. Do you have mongo records per hosted customer where you can store this?

I know storing state isn't desired but I think Tristan's recent post about the line between what's open source and what's paid for is a good model. If its paid for then some state can be stored.

**Reply by TuringLovesDeathMetal** [reply]
ooh good question! 

So, this would mean:
- we store the state of the repo in Lightdash (I think we do this already? 🤔 otherwise we'd refresh dbt every time you refreshed the page I think? - I'm not sure if we do this actually...)
- We have the name of the commit that this state is coming from 
- When you open Lightdash/login on startup, we check the commit name for the state, then we check GitHub to see if there is a newer commit. 
- If there's no newer commit, you just crack on. 
- If there is a newer commit, we refresh dbt for you in the background.

**Reply by djayatillake** [reply]
Yeah exactly I don't think you store state of repo currently as there is always a lag on login where the repo state is pulled.

**Comment by TuringLovesDeathMetal** [comment]
Some more feedback on this here: https://getdbt.slack.com/archives/C026WJE4A69/p1642603234013700

I think there are a few key things here:
1. Can we/do we store the state of the current repo so that booting up Lightdash is faster? 
2. We could read your `manifest.json` in Lightdash (instead of building it for you)...This would only really be useful if it's connected to GitHub and the `manifest.json` get compiled + sent to Lightdash when there's a new merge (so you don't have to refresh dbt yourself, manually). 
3. ^We could add to this GitHub action so that we check if your Lightdash project has any errors (e.g. metrics are not defined properly) before you merge the changes and sync them in Lightdash. 
4. We could change the `refresh dbt` into a page which has information about your connection and errors so that it makes debugging easier.

None of these feel like _**huge**_ step changes in the product (except maybe 3...) and they're quite a bit of work. 

I think it would be useful to get some more users into Lightdash and see what their feedback is before we make any moves on this...
